add workspace trust required editor

sbatten opened the pull request:
Sorry for the spam review requests. This is the basic logic to do what we discussed today. An `EditorInput` can determine if it needs trust to be opened correctly and if trust is not given, the editor control will use a fallback EditorPane. We also reopen on trust transitions automatically.

The text and buttons in the fallback editor can be changed and reviewed after the base functionality.

lramos15 commented:
Some things I noticed while testing with all custom editors set to true

1. The trust editor pane probably shouldn't be marked as dirty, as technically things like undo and save won't work and there was cases where doing these actions would produce dirty indicator changes, but actually results in data loss. I'm not sure how you can tackle this as you aren't touching the model, just the pane.
2. You can't test enabling / disabling trust as I run into what @bpasero was talking about involving custom editors failing to restore when the ext host restarts. This isn't your fault, but definitely something that needs to be looked into

Looks to work rather well with restoring though.

bpasero commented on src/vs/workbench/common/editor.ts:
```diff
@@ -68,6 +68,11 @@ export const TEXT_DIFF_EDITOR_ID = 'workbench.editors.textDiffEditor';
+/**
+ * Workspace Trust Required editor id.
+ */
+export const WORKSPACE_TRUST_REQUIRED_FILE_EDITOR_ID = 'workbench.editors.files.workspaceTrustRequiredEditor';
```
```suggestion
export const WORKSPACE_TRUST_REQUIRED_FILE_EDITOR_ID = 'workbench.editors.workspaceTrustRequiredEditor';
```

bpasero commented on src/vs/workbench/common/editor.ts:
```diff
@@ -462,6 +467,11 @@ export interface IEditorInput extends IDisposable {
 	resolve(): Promise<IEditorModel | null>;
 
+	/**
+	 * Returns if the input requires trust or not.
```
```suggestion
	 * Returns if the input requires workspace trust or not.
```

bpasero commented on src/vs/workbench/common/editor.ts:
```diff
@@ -462,6 +467,11 @@ export interface IEditorInput extends IDisposable {
+	/**
+	 * Returns if the input requires trust or not.
+	 */
+	requiresTrust(): Promise<boolean>;
```
Should this be `requiresWorkspaceTrust`?

bpasero commented on src/vs/workbench/common/editor.ts:
```diff
@@ -602,6 +612,10 @@ export abstract class EditorInput extends Disposable implements IEditorInput {
 		return { typeId: this.typeId };
 	}
 
+	async requiresTrust(): Promise<boolean> {
```
Please also provide a default implementation for the `SideBySideEditor` below which should return `true` if any of the two editors inside require trust.

bpasero commented on src/vs/workbench/browser/parts/editor/editorControl.ts:
```diff
@@ -48,15 +57,29 @@ export class EditorControl extends Disposable {
+		const blockedByTrust = requiresTrust && !this.workspaceTrustService.isWorkpaceTrusted();
+
+		if (requiresTrust) {
+			const disposable = this.workspaceTrustService.onDidChangeTrust(() => {
```
Not so happy about accumulating a listener each time an editor opens, wouldn't it be better to listen to trust changes from the constructor and then update the active pane accordingly if it is a workspace trust editor?

bpasero commented on src/vs/workbench/browser/parts/editor/editorControl.ts:
```diff
@@ -48,15 +57,29 @@ export class EditorControl extends Disposable {
+		if (requiresTrust) {
+			const disposable = this.workspaceTrustService.onDidChangeTrust(() => {
+				if (this._activeEditorPane?.input === editor) {
+					this.openEditor(editor, options, context);
```
I would use `matches` instead of `===` to be safe.

bpasero reviewed (changes requested):
Inside `editorControl` there is not only `openEditor` but also [`closeEditor`](https://github.com/microsoft/vscode/blob/190e902afe6132b25714750035df13efd94ab409/src/vs/workbench/browser/parts/editor/editorControl.ts#L241-L241) and we need to double check that this works even when the trust editor is visible.

sbatten commented:
@bpasero addressed all the feedbacks but I don't understand the top-level comment of your review.
![recording (5)](https://user-images.githubusercontent.com/6561887/117475647-29788080-af11-11eb-87e2-56a4f92fd789.gif)

bpasero reviewed (changes requested):
@sbatten I pushed some :lipstick: and more importantly a [change](https://github.com/microsoft/vscode/pull/123181/commits/131bc5a251b7cfde5de80bc3ed026282b5fb7fe4) to go through the proper `openEditor` method when reopening from trust change to get the same behaviour as a normal editor open would get (e.g. how errors are handled).

I am ok to 🚢  this after you added a test case. I think it should be possible to test the flow (first untrusted, then trusted) quite easily. I think a good location would be `src/vs/workbench/test/browser/parts/editor/editorPane.test.ts`.
